src/dhcp6relay/src/relay.cpp

```diff
@@ -0,0 +1,665 @@
+		syslog(LOG_ERR, "bind: Failed to bind to specified interface\n");
+		(void) close(s);
+	    return -1;
+	}
+
+	struct packet_mreq mreq = {
+	    .mr_ifindex = ifindex,
+	    .mr_type = PACKET_MR_PROMISC
+	};
+
+	if (setsockopt(s, SOL_PACKET, PACKET_ADD_MEMBERSHIP, &mreq, sizeof mreq) == -1)
+	{
+		syslog(LOG_ERR, "setsockopt: Failed to send PACkET_MR_PROMISC\n");
+		(void) close(s);
+	    return -1;
+	}
+
+	if (fprog && setsockopt(s, SOL_SOCKET, SO_ATTACH_FILTER, fprog, sizeof *fprog) == -1)
+	{
+		syslog(LOG_ERR, "setsockopt: Failed to attach filter\n");
+		(void) close(s);
+	    return -1;
+	}
+
+	return s;
+}
+
+/**
+ * @code                        prepare_relay_config(relay_config *interface_config, int fd, const arg_config *context);
+ * 
+ * @brief                       prepare for specified relay interface config: server and link address
+ *
+ * @param interface_config      pointer to relay config to be prepared
+ * @param local_sock            L3 socket used for relaying messages
+ * @param filter                socket attached with filter
+ * @param config                argument config that contains strings of server and interface addresses
+ *
+ * @return                      none
+ */
+void prepare_relay_config(relay_config *interface_config, int local_sock, int filter, const arg_config *context) {
+    interface_config->local_sock = local_sock; 
+    interface_config->filter = filter; 
+    interface_config->is_option_79 = context->is_option_79; 
+
+    for(auto server: context->servers) {
+        sockaddr_in6 tmp;
+        if(inet_pton(AF_INET6, server.c_str(), &tmp.sin6_addr) != 1)
+        {
+            syslog(LOG_WARNING, "inet_pton: Failed to convert IPv6 address\n");
+        }
+        tmp.sin6_family = AF_INET6;
+        tmp.sin6_flowinfo = 0;
+        tmp.sin6_port = htons(547);
+        tmp.sin6_scope_id = 0; 
+        interface_config->servers.push_back(tmp);
+    }
+
+    sockaddr_in6 *non_link_local = NULL;    
+    sockaddr_in6 *link_local = NULL;
+    
+    struct ifaddrs *ifa, *ifa_tmp;
+    char addr[INET6_ADDRSTRLEN];
+    addr[0] = 0;
+
+    if (getifaddrs(&ifa) == -1) {
+		    syslog(LOG_WARNING, "getifaddrs: Unable to get network interfaces\n");
+            exit(1);
+	}
+
+    ifa_tmp = ifa;
+	while (ifa_tmp) {
+		if ((ifa_tmp->ifa_addr) && (ifa_tmp->ifa_addr->sa_family == AF_INET6)) {
+            struct sockaddr_in6 *in6 = (struct sockaddr_in6*) ifa_tmp->ifa_addr;
+            inet_ntop(AF_INET6, &in6->sin6_addr, addr, sizeof(addr));
+            if((strcmp(ifa_tmp->ifa_name, interface_config->interface.c_str()) == 0) && is_addr_gua(in6->sin6_addr)) {    
+                non_link_local = in6;
+            }
+            if((strcmp(ifa_tmp->ifa_name, interface_config->interface.c_str()) == 0) && is_addr_link_local(in6->sin6_addr)) {    
+                link_local = in6;
+            }   
+	    }
+	    ifa_tmp = ifa_tmp->ifa_next;
+	}
+	freeifaddrs(ifa);
+
+    if (non_link_local != NULL) {
```

Comment:
what about multiple link local and/or global addresses?

Reply:
Any global address will do. Link local address is only used if not a single global address is present